### PR TITLE
Sim/`DexMoves`: Deduplicate `DataMove` against parent mod

### PR DESCRIPTION
**Pull request by larry-the-table-guy**
Dedup for `DataMove`.
	
`ret size`: Retained Size column in Chrome dev tools memory tab. "Size of the object plus the graph it retains"

| stage | `DexMoves` ret size | `DataMove` ret size | `allMods.moves.all()` | total heap size |
| --- | ---- | ----- | --- | --- |
| Baseline | 24.345 MB | 22.747 MB | 475 ms | 260 MB |
| dedup moves | 3.3 MB | 3.0 MB | 402 ms | 240 MB |

36344 -> 4968 instances of `DataMove`.

I was really shocked to find that it loads faster - it's running **more** code.
I'm guessing the 86% dedup rate means that so much less memory (which now never leaves the 'young gen') had to be copied by the GC, so maybe that's where the time savings come from? 
The first couple paragraphs of [this v8 blogpost](https://v8.dev/blog/orinoco-parallel-scavenger) explain the role of the young gen in GCs.

[Earlier tests](https://github.com/smogon/pokemon-showdown/pull/10549#issuecomment-2407772806) indicate that this strategy can, at best, yield 4739 instances of `DataMove`, and Chrome's UI indicates that there are 4968 instances of DataMove. That test was over a month ago, so maybe some mods changed? Anyway, we're in the right ballpark. Oh, the other cause could be overuse of `modData`, which does a deep copy. The earlier test uses deep equality but this impl uses shallow equality (much cheaper, but has false negatives (safe for correctness)). Maybe there's some script that calls `modData` but doesn't actually modify the move. Not a big deal atm.

---

You can save a further 15% in time (and <1 MB) by also doing this:

```ts
all(): readonly Move[] {
        if (this.allCache) return this.allCache;
        // BEGIN NEW
        if (this.dex.parentMod) {
                const parent = this.dex.mod(this.dex.parentMod);
                if (this.dex.data.Moves === parent.data.Moves && this.dex.gen === parent.gen) {
                        this.allCache = parent.moves.all();
                        this.moveCache = parent.moves.moveCache;
                        return this.allCache;
                }
        }
        // END NEW
        const moves = [];
        for (const id in this.dex.data.Moves) {
                moves.push(this.getByID(id as ID));
        }
        this.allCache = Object.freeze(moves);
        return this.allCache;
}
```

I left it out bc imo this is a better fit for when all objects are constructed during DexMoves constructor, but that requires lazy mod loading. Constructing all the objects upfront in DexMoves' ctor keeps the logic in one place, and is likely to be faster.

---

<details><summary>bench code</summary>

Inserted at the bottom of `sim/dex.ts`

```ts
// const v8 = require('v8'); // uncomment the two v8 lines to do a heap snapshot which you can inspect with Chrome.
Dex.includeMods();
Dex.includeModData();
for (const e of ['abilities', 'species', 'moves', 'items', 'types', 'natures']) {
        const start = performance.now();
        for (const d in dexes) {
                if (d === 'base') continue;
                const m = dexes[d];
                m[e].all();
        }
        console.log(e, performance.now() - start);
}
// v8.writeHeapSnapshot();
throw new Error('stop early');
```

</details>

---

<details><summary>load time for all other Dex component</summary>

Note: in a similar PR I left in the `v8.writeHeapSnapshot` calls when measuring the time 
for `.all()`s. For whatever reason, that noticeably slows down the `.all()`s. So, that's
why these numbers add up to a much smaller time than the total time recorded in the other PR.

Time in milliseconds.
These times were consistent with repeated runs.

```
// without dedup
abilities 96.11295000091195
species 713.9901599995792
moves 475.5642639994621
items 132.09175600111485
types 1.7992599979043007
natures 2.300714001059532

// with dedup (this PR)
abilities 96.6237739995122
species 716.0696760006249
moves 402.00951999798417
items 131.15769999846816
types 1.6965250000357628
natures 2.189902000129223


// also with same-gen dedup in .all() 
abilities 95.16279899701476
species 700.7795180007815
moves 343.8259319998324
items 127.67007499933243
types 1.6952290013432503
natures 2.1241800002753735
```

</details>

### CODE REVIEW

**Comment by larry-the-table-guy** [comment]
Ah, a simpler explanation for the decrease in time is that `dex.deepFreeze` will exit early on 86% of the moves (bc they're already frozen).

**Comment by larry-the-table-guy** [comment]
Rearranged the check so that the parent move isn't fetched unless the data matches (and thus there's a very high chance of sharing the object). Better for cases where a process only loads a few mods.

Below benchmark went from ~80ms to ~60ms
```ts
const gen1 = Dex.mod('gen1');
const start = performance.now();
gen1.moves.all();
console.log(performance.now() - start);
throw new Error('stop early');
```
Note that the `loadData` chain happens during `Dex.mod()`

**Comment by DaWoblefet** [comment]
Just to make sure, this is only memory-related, right? We commonly have differences in move data that aren't just ``isNonstandard`` properties, and which don't get modified in descriptions. Differences in move flags from generation to generation come to mind.

**Comment by larry-the-table-guy** [comment]
> Just to make sure, this is only memory-related, right? We commonly have differences in move data that aren't just `isNonstandard` properties, and which don't get modified in descriptions. Differences in move flags from generation to generation come to mind.

Right, the primary motivation is to reduce memory usage by sharing identical objects between mods. The intent is for the objects to be shared only if every value is identical. My reasoning is that if the inputs to the constructor are the same (`moveData`, `desc`, `shortDesc`), and all the fields that get modified outside the constructor (`isNonstandard`) are the same, then the objects will be the same, and that remains true post-construction since they are then deep-frozen.

It could also be implemented with a `deepEquals` function, but that's a fair bit slower. Not catastrophically slower, granted. I can make a branch that also does that to double check that the objects are indeed identical.

**Comment by larry-the-table-guy** [comment]
Ok, so adding this line:
```ts
            if (this.dex.parentMod) {
                    const parentMod = this.dex.mod(this.dex.parentMod);
                    if (moveData === parentMod.data.Moves[id]) {
                            const parentMove = parentMod.moves.getByID(id);
                            if (move.isNonstandard === parentMove.isNonstandard &&
                                 move.desc === parentMove.desc &&
                                 move.shortDesc === parentMove.shortDesc) {
                                     require('assert').deepStrictEqual(move, parentMove); // NEW
                                     move = parentMove;
                            }
                    }
            }
```
Runs without error when loading all the mods.
Time to construct all moves for all mods (excluding time for `loadData`), went from 400ms to 800ms. Not that bad in the grand scheme of things.

(You could wrap the assert in a try-catch so that it just aborts the dedup instead of crashing)

File: sim/dex-moves.ts [hunk 1]
```diff
@@ -655,6 +655,17 @@ export class DexMoves {
 			if (move.gen > this.dex.gen) {
 				(move as any).isNonstandard = 'Future';
 			}
+			if (this.dex.parentMod) {
+				const parentMod = this.dex.mod(this.dex.parentMod);
```
Comment (Zarel) [inline sim/dex-moves.ts:659]:
```suggestion
				// if move is exactly identical to parentMod's move, just use the old copy
				const parentMod = this.dex.mod(this.dex.parentMod);
```

Should probably have a comment so this is clear that's what's going on.

**Review by Zarel — APPROVED** [review]
Approved, other than I think needs a comment.

**Comment by Zarel** [comment]
Thanks!

**Pull request merged**